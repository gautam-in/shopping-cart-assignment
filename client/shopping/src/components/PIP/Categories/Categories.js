import React, { useState } from 'react';

function Categories({categories}) {
    const [state, setState] = useState({toggle:false, current: categories[0].name})
    let dropDown = ()=>{
        setState({...state, toggle:!state.toggle})
    }
    let setCategory = (data)=>{
        setState({toggle:!state.toggle, current:data})
    }
    return (
        <>
        <div className="row">
            <div className="col-12 categoriesTab mobiledropdown" style={{order:0}} onClick={dropDown}>
                <div style={{display:'flex', justifyContent:"space-between"}}><div>{state.current}</div> <div>V</div></div>
            </div>
            <div className={state.toggle?"open":"close"}>
                {Object.keys(categories).map((key)=>{
                    return categories[key].order>=0 && <div className="col-11 categoriesTab" style={{order:categories[key].order}} onClick={()=>setCategory(categories[key].name)}>
                        {categories[key].name}
                    </div>
                })} 
            </div>
        </div>
        </>
    );
}

export default Categories;