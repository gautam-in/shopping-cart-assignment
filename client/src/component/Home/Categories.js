import CategoryItem from './CategoryItem';
import { useEffect, useState } from 'react';
import axios from 'axios';

import { CategoriesContainer } from './StyledHome';

const Categories = () => {
	const [categoryItems, setCategoryItems] = useState([]);

	useEffect(() => {
		axios
			.get('http://localhost:5000/categories')
			.then((res) => res.data)
			.then((result) => {
				setCategoryItems(result);
			});
	}, []);

	return (
		<CategoriesContainer>
			{categoryItems.map((categoryItem) =>
				categoryItem.enabled ? (
					<CategoryItem item={categoryItem} key={categoryItem.id} />
				) : null
			)}
		</CategoriesContainer>
	);
};

export default Categories;
