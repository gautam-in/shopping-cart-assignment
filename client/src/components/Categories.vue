<template>
  <section id="categoryContainer">
    <div id="bannerContainer" class="container">
      <div class="row">
        <div class="categoryCol col-lg-12 col-xl-12 col-md-12 col-sm-12 col-12">
          <div
            id="carouselExampleIndicators"
            class="carousel carousel-dark slide"
            data-bs-ride="carousel"
          >
            <div class="carousel-indicators">
              <button
                v-for="(item, index) in banners"
                :key="index"
                type="button"
                :data-bs-target="'#carouselExampleIndicators'"
                :data-bs-slide-to="index"
                :class="index == 0 ? 'active' : ''"
                :aria-current="index == 0 ? true : false"
                :aria-label="'Slide' + index"
              ></button>
            </div>
            <div class="carousel-inner">
              <div
                class="carousel-item"
                :class="index == 0 ? 'active' : ''"
                v-for="(item, index) in banners"
                :key="index"
              >
                <img
                  :src="require('@/assets' + item.bannerImageUrl)"
                  class="d-block w-100"
                  :alt="item.bannerImageAlt"
                />
              </div>
            </div>
            <button
              class="carousel-control-prev"
              type="button"
              data-bs-target="#carouselExampleIndicators"
              data-bs-slide="prev"
            >
              <span
                class="carousel-control-prev-icon"
                aria-hidden="true"
              ></span>
              <span class="visually-hidden">Previous</span>
            </button>
            <button
              class="carousel-control-next"
              type="button"
              data-bs-target="#carouselExampleIndicators"
              data-bs-slide="next"
            >
              <span
                class="carousel-control-next-icon"
                aria-hidden="true"
              ></span>
              <span class="visually-hidden">Next</span>
            </button>
          </div>
        </div>
        <div
          v-for="(category, index) in categories"
          :key="index"
          :id="'category' + index"
          class="categoryCol col-lg-12 col-md-12 col-xl-12 col-sm-12 col-12"
        >
          <div class="categoryListing row">
            <div id=""></div>
            <div id=""></div>
          </div>
        </div>
      </div>
    </div>
  </section>
</template>

<script>
export default {
  name: "Categories",
  props: ["banners", "categories"],
  mounted() {
    this.categories.forEach((val, ind) => {
      if (val.imageUrl) {
        let categoryContainer = document.querySelector("#category" + ind);
        let img = document.createElement("img");

        img.src = require("@/assets" + val.imageUrl);
        // img.style.height = "200px";
        // img.style.width = "300px";
        img.style.padding = "10px";
        img.classList.add("imageSizing");
        img.alt = val.bannerImageAlt;

        let p = document.createElement("p");
        let p1 = document.createElement("p");

        p.innerText = val.name;
        p.style.fontWeight = "600";

        p1.innerText = val.description;
        p1.style.fontSize = "12px";

        if (ind % 2 == 0) {
          categoryContainer.firstElementChild.firstChild.classList.add(
            "col-lg-6",
            "col-xl-6",
            "col-md-6",
            "col-sm-4",
            "col-4"
          );
          categoryContainer.firstElementChild.lastChild.classList.add(
            "col-lg-6",
            "col-xl-6",
            "col-md-6",
            "col-sm-8",
            "col-8"
          );
          categoryContainer.firstElementChild.firstChild.append(img);

          let info = document.createElement("div");
          let btn = document.createElement("button");
          btn.value = "Explore " + val.key;
          btn.innerText = "Explore " + val.key;
          btn.classList.add("btnStyle");
          info.classList.add("infoPadding");
          categoryContainer.firstElementChild.lastChild.append(info);
          info.append(p);
          info.append(p1);
          info.append(btn);
        } else {
          categoryContainer.firstElementChild.lastChild.classList.add(
            "col-lg-6",
            "col-xl-6",
            "col-md-6",
            "col-sm-4",
            "col-4"
          );
          categoryContainer.firstElementChild.firstChild.classList.add(
            "col-lg-6",
            "col-xl-6",
            "col-md-6",
            "col-sm-8",
            "col-8"
          );
          categoryContainer.firstElementChild.lastChild.append(img);
          let info = document.createElement("div");
          let btn = document.createElement("button");
          btn.value = "Explore " + val.key;
          btn.innerText = "Explore " + val.key;
          btn.classList.add("btnStyle");
          info.classList.add("infoPadding");
          categoryContainer.firstElementChild.firstChild.append(info);
          info.append(p);
          info.append(p1);
          info.append(btn);
        }
      }
    });
  },
};
</script>

<!-- Add "scoped" attribute to limit CSS to this component only -->
<style lang="scss" scoped>
.row > .categoryCol {
  box-shadow: 0 6px 7px -4px rgb(0 0 0 / 40%);
  margin-bottom: 10px;
}
</style>
