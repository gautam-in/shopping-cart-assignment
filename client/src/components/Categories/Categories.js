import React, { useEffect, useState } from "react";
import Category from "./Category";
import { fetchCategoriesAction } from "../../Action";
import { store } from "../../store";
import { shallowEqual, useSelector } from "react-redux";
const categoriesStore = (state) => state.categories.map((category) => category);

function Categories() {
  const [categories, setCategories] = useState([]);
  // const [categoriesEle, setcategoriesEle] = useState([]);
  const categoriesSelect = useSelector(categoriesStore, shallowEqual);

  useEffect(() => {
    store.dispatch(fetchCategoriesAction());
  }, []);
  useEffect(() => {
    const getCat = categoriesSelect.map((category, index) => (
      <Category category={category} index={index} key={category.id} />
    ));
    setCategories(getCat);
  }, [categoriesSelect]);
  return <div className="container">{categories}</div>;
}

export default Categories;
