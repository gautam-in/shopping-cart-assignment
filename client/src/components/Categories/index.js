import React, { useEffect, useState } from "react";
import { fetchCategories } from "../../api/products";
import CategoryCard from "../CategoryCard";

function Categories() {
  const [categories, setCategories] = useState([]);

  useEffect(() => {
    const controller = new AbortController();

    fetchCategories({ signal: controller.signal }).then((categories) => {
      setCategories(categories);
    });

    return () => controller.abort();
  }, []);

  return (
    <>
      {categories.map(({ id, key, name, imageUrl, description }, index) => (
        <CategoryCard
          id={id}
          key={id}
          slug={key}
          name={name}
          imageUrl={imageUrl}
          description={description}
        />
      ))}
    </>
  );
}

export default Categories;
