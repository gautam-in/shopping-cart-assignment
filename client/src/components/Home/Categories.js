// react
import React from 'react';
import PropTypes from 'prop-types';
import { useNavigate } from 'react-router-dom';

/**
 * @name Categories
 * @param {categoriesData} param0
 * @returns JSX for Categories Section
 */
const Categories = ({ categoriesData }) => {
  const navigate = useNavigate();

  return (
    <div className="categories">
      {categoriesData?.map((category, index) => {
        const { id, name, description, key, imageUrl } = category;
        return (
          <div
            key={id}
            className={(index + 1) % 2 === 0 ? 'category img-left' : 'category img-right'}>
            <div className="category__body">
              <h2 className="category__body__name">{name}</h2>
              <p className="category__body__description">{description}</p>
              <button
                className="category__body__button"
                onClick={() => navigate(`/products/${name}`)}>{`Explore ${key}`}</button>
            </div>

            <div className="category__image">
              <img src={imageUrl} alt={name} />
            </div>
          </div>
        );
      })}
    </div>
  );
};

Categories.propTypes = {
  categoriesData: PropTypes.array
};

export default Categories;
