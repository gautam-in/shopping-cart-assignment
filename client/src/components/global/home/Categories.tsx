import { useMemo } from 'react';
import { useNavigate } from 'react-router-dom';
import useQuery from '../../../hooks/useQuery';
import { Category } from '../../../utils/types/category';
import CategoryItem from './CategoryItem';

function Categories() {
  const { data: categories } = useQuery('/categories');
  const navigate = useNavigate();

  const categoryOrder = useMemo(
    () =>
      categories
        .sort((a: Category, b: Category) => a.order - b.order)
        .filter((item: Category) => item.key !== 'seafood'),
    [categories],
  );

  const handleCategoryClick = (category: Category) => {
    navigate(`/products`, { state: category.id });
  };

  return (
    <section className="categories">
      {categoryOrder.map((category: Category, index: number) => (
        <CategoryItem
          category={category}
          index={index}
          key={category.id}
          onClick={() => handleCategoryClick(category)}
        />
      ))}
    </section>
  );
}

export default Categories;
