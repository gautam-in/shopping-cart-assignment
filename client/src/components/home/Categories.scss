@use "../common/base";

.item {
	width: 80%;
	margin: auto;
	border-bottom: base.$gray-color;
	box-shadow: 0 8px 6px -6px rgba(138, 142, 143, 0.5);
	text-align: center;
	font-size: 1em;
	display: grid;
	grid-template-columns: repeat(3, 1fr);
	padding: 2em;
	img {
		width: 30vw;
		height: 10rem;
		grid-column: 1/2;
	}
	section {
		grid-column: 2/-1;
		display: flex;
		flex-direction: column;
		align-items: center;
		justify-content: center;

		.explore {
			text-decoration: none;
			margin-bottom: 1rem;
			box-shadow: 0 0.1em 0.1em rgb(216, 210, 210);
			background-color: base.$pink-color;
			color: white;
			padding: 0.2em 0.5em;
			border: none;
		}
	}
}

.categories-items :nth-child(even) {
	img {
		order: 1;
		grid-column: 3/-1;
	}
	section {
		grid-column: 1/3;
	}
}

@media screen and (max-width: 520px) {
	.item {
		font-size: 0.6em;
		img {
			width: 8em;
			height: 7em;
		}
		h1 {
			font-size: 0.8em;
		}
		p {
			font-size: 0.6em;
		}
		button: {
			font-size: 0.5em;
		}
	}
}
