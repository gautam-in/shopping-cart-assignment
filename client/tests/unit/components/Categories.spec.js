import { mount } from "@vue/test-utils";
import Categories from "@/components/Categories.vue";
import Vue from "vue";
import VueRouter from "vue-router";

describe("Categories", () => {
  it("snapshot", () => {
    const wrapper = mount(Categories, {
      propsData: {
        banners: [],
        categories: [],
      },
    });
    expect(wrapper).toMatchSnapshot();

    // expect(wrapper.contains("div")).toBe(true);
  });
});
