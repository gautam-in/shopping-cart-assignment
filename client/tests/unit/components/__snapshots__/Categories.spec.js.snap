// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Categories snapshot 1`] = `
<section id="categoryContainer">
  <div id="bannerContainer" class="container">
    <div class="row">
      <div class="categoryCol col-lg-12 col-xl-12 col-md-12 col-sm-12 col-12">
        <div id="carouselExampleIndicators" data-bs-ride="carousel" class="carousel carousel-dark slide">
          <div class="carousel-indicators"></div>
          <div class="carousel-inner"></div> <button type="button" data-bs-target="#carouselExampleIndicators" data-bs-slide="prev" class="carousel-control-prev"><span aria-hidden="true" class="carousel-control-prev-icon"></span> <span class="visually-hidden">Previous</span></button> <button type="button" data-bs-target="#carouselExampleIndicators" data-bs-slide="next" class="carousel-control-next"><span aria-hidden="true" class="carousel-control-next-icon"></span> <span class="visually-hidden">Next</span></button>
        </div>
      </div>
    </div>
  </div>
</section>
`;
